File: src/test/java/main/QuetionsAndAnswers/CategoriesTests.java
package main.QuetionsAndAnswers;
import main.ApplicationsNavigationTest;
import main.PublicTenderPage;
import main.TendersPage;
import main.questionAndAnswers.Answers;
import main.questionAndAnswers.Categories;
import org.openqa.selenium.Alert;
import org.testng.annotations.Test;

public class CategoriesTests extends ApplicationsNavigationTest {

    @Test
    public void addNewCategotyforQuestions() throws InterruptedException {
        TendersPage tendersPage = new TendersPage(driver);
        tendersPage.switchToBrowserFrame();
        PublicTenderPage publicTenderPage = new PublicTenderPage(driver);
        publicTenderPage.clickOnElement(publicTenderPage.getQuestionsAndAnswersTab());
        driver.switchTo().defaultContent();
        tendersPage.switchToNavigationFrame();
        Categories categories = new Categories(driver);
        String mainWindowHandle = driver.getWindowHandle();
        categories.clickManageCategoriesButton();
        tendersPage.switchToNewWindowTest();
        categories.enterCategoryName("reTest 1");
        categories.clickOkButton();
        driver.switchTo().window(mainWindowHandle);
        tendersPage.switchToNavigationFrame();
        Thread.sleep(3000);
        categories.clickManageCategoriesButton();
        Thread.sleep(3000);
        tendersPage.switchToNewWindowTest();
        categories.enterCategoryName("to Delete");
        categories.clickOkButton();
        driver.switchTo().window(mainWindowHandle);
        tendersPage.switchToNavigationFrame();
        Thread.sleep(3000);
        categories.clickManageCategoriesButton();
        Thread.sleep(3000);
        tendersPage.switchToNewWindowTest();
        categories.enterCategoryName("Test Categoty 2");
        categories.clickOkButton();
    }
    @Test
    public void deleteCategory() throws InterruptedException {
        TendersPage tendersPage = new TendersPage(driver);
        tendersPage.switchToBrowserFrame();
        PublicTenderPage publicTenderPage = new PublicTenderPage(driver);
        publicTenderPage.clickOnElement(publicTenderPage.getQuestionsAndAnswersTab());
        driver.switchTo().defaultContent();
        tendersPage.switchToNavigationFrame();
        Categories categories = new Categories(driver);
        categories.clickManageCategoriesButton();
        tendersPage.switchToNewWindowTest();
        categories.selectCategoryToDelete();
        categories.clickDelete();
        Alert alert = driver.switchTo().alert();
        alert.accept();
        Thread.sleep(3000);}
@Test
    public void deleteCategoryInUse() throws InterruptedException {
    TendersPage tendersPage = new TendersPage(driver);
    tendersPage.switchToBrowserFrame();
    PublicTenderPage publicTenderPage = new PublicTenderPage(driver);
    publicTenderPage.clickOnElement(publicTenderPage.getQuestionsAndAnswersTab());
    driver.switchTo().defaultContent();
    tendersPage.switchToNavigationFrame();
    Categories categories = new Categories(driver);
    categories.clickManageCategoriesButton();
    tendersPage.switchToNewWindowTest();
    categories.selectCategoryInUse();
    categories.clickDelete();
    Alert alert = driver.switchTo().alert();
    alert.accept();
    Thread.sleep(3000);
}
@Test
public void editCategoryNameInPublishedAnswers() throws InterruptedException {
    TendersPage tendersPage = new TendersPage(driver);
    tendersPage.switchToBrowserFrame();
    PublicTenderPage publicTenderPage = new PublicTenderPage(driver);
    publicTenderPage.clickOnElement(publicTenderPage.getQuestionsAndAnswersTab());
    driver.switchTo().defaultContent();
    tendersPage.switchToNavigationFrame();
    Categories categories = new Categories(driver);
    Answers answers = new Answers(driver);
    driver.switchTo().defaultContent();
    tendersPage.switchToDirectoryFrame();
    answers.goToPublishedAnswersTab();
    Thread.sleep(3000);
    answers.selectQuestionToAnswer();
    categories.clickEditCategoryButton();
    tendersPage.switchToNewWindowTest();
    categories.enterCategoryName("edited category name");
    categories.clickOkButton();
//        Alert alert = driver.switchTo().alert();
//        alert.accept();
    Thread.sleep(3000);

}
    @Test
    public void deleteCategoryWithPublishedAnswers() throws InterruptedException {
        TendersPage tendersPage = new TendersPage(driver);
        tendersPage.switchToBrowserFrame();
        PublicTenderPage publicTenderPage = new PublicTenderPage(driver);
        publicTenderPage.clickOnElement(publicTenderPage.getQuestionsAndAnswersTab());
        driver.switchTo().defaultContent();
        tendersPage.switchToNavigationFrame();
        Categories categories = new Categories(driver);
        Answers answers = new Answers(driver);
        driver.switchTo().defaultContent();
        tendersPage.switchToDirectoryFrame();
        answers.goToPublishedAnswersTab();
        Thread.sleep(3000);
        answers.selectQuestionToAnswer();
        categories.clickEditCategoryButton();
        tendersPage.switchToNewWindowTest();
        categories.clickDelete();
        categories.clickOkButton();
        Alert alert = driver.switchTo().alert();
        alert.accept();
        Thread.sleep(3000);

    }

}

File: src/test/java/main/QuetionsAndAnswers/CategoriesTests.kt
package main.QuetionsAndAnswers

import main.ApplicationsNavigationTest
import main.BaseTest
import main.PublicTenderPage
import main.TendersPage
import main.questionAndAnswers.Answers
import main.questionAndAnswers.Categories
import org.testng.annotations.BeforeMethod
import org.testng.annotations.Test

class CategoriesTests : ApplicationsNavigationTest() {
    override  var tendersPage: TendersPage? = null
    private lateinit var publicTenderPage: PublicTenderPage
    private lateinit var categories: Categories
    private lateinit var answers: Answers
    @BeforeMethod
    fun setUpTest() {
        tendersPage = TendersPage(driver)
        tendersPage!!.switchToBrowserFrame()

        publicTenderPage = PublicTenderPage(driver)
        categories = Categories(driver)
        answers = Answers(driver)
        publicTenderPage.questionsAndAnswersTab?.click()
        driver.switchTo().defaultContent()
        tendersPage!!.switchToNavigationFrame()
    }

    @Test
    fun addNewCategoryForQuestions() {
        driver.findElement(categories.manageCategoriesButton).click()
        tendersPage?.switchToNewWindowTest()
        categories.enterCategoryName("Test 1")
        driver.findElement(categories.okButton).click()
        Thread.sleep(3000)
    }

    @Test
    fun addCategoryToDelete() {
        driver.findElement(categories.manageCategoriesButton).click()
        tendersPage?.switchToNewWindowTest()
        categories.enterCategoryName("to Delete")
        driver.findElement(categories.okButton).click()
        Thread.sleep(3000)
    }

    @Test
    fun addCategory2() {
        driver.findElement(categories.manageCategoriesButton).click()
        Thread.sleep(3000)
        tendersPage?.switchToNewWindowTest()
        categories.enterCategoryName("Test Category 2")
        driver.findElement(categories.okButton).click()
        Thread.sleep(3000)
    }

    @Test
    fun addCategoryTooChange() {
        driver.findElement(categories.manageCategoriesButton).click()
        Thread.sleep(3000)
        tendersPage?.switchToNewWindowTest()
        categories.enterCategoryName("category TO Change")
        driver.findElement(categories.okButton).click()
    }

    @Test
    fun deleteCategory() {
        driver.findElement(categories.manageCategoriesButton).click()
        tendersPage?.switchToNewWindowTest()
        categories.selectCategoryToDelete()
        driver.findElement(categories.deleteCategoryButton).click()
        val alert = driver.switchTo().alert()
        alert.accept()
        Thread.sleep(3000)
    }

    @Test
    fun deleteCategoryInUse() {
        driver.findElement(categories.manageCategoriesButton).click()
        tendersPage?.switchToNewWindowTest()
        categories.selectCategoryInUse()
        driver.findElement(categories.deleteCategoryButton).click()
        val alert = driver.switchTo().alert()
        alert.accept()
        Thread.sleep(3000)
    }

    @Test
    fun editCategoryNameInPublishedAnswers() {
        driver.switchTo().defaultContent()
        tendersPage?.switchToDirectoryFrame()
        answers.goToPublishedAnswersTab()
        Thread.sleep(3000)
        answers.selectQuestionToAnswer()
        driver.findElement(categories.editCategoryButton).click()
        tendersPage?.switchToNewWindowTest()
        categories.enterCategoryName("edited category name")
        driver.findElement(categories.okButton).click()
        Thread.sleep(3000)
    }

    @Test
    fun deleteCategoryWithPublishedAnswers() {
        driver.switchTo().defaultContent()
        tendersPage?.switchToDirectoryFrame()
        answers.goToPublishedAnswersTab()
        Thread.sleep(3000)
        answers.selectQuestionToAnswer()
        driver.findElement(categories.editCategoryButton).click()
        tendersPage?.switchToNewWindowTest()
        driver.findElement(categories.deleteCategoryButton).click()
        val alert = driver.switchTo().alert()
        alert.accept()
        Thread.sleep(3000)
    }

    @Test
    fun changeCategoryForPublishedAnswers() {
        driver.switchTo().defaultContent()
        tendersPage?.switchToDirectoryFrame()
        answers.goToPublishedAnswersTab()
        Thread.sleep(3000)
        answers.selectQuestionToAnswer()

        driver.findElement(categories.changeCategoryButton).click()
        tendersPage?.switchToNewWindowTest()
        categories.changeCategoryForPublishedAnswer()
        driver.findElement(categories.okButton).click()
    }

}
//TODO review all test, some buttons are not working
